Framework/API/inc/MantidAPI/Algorithm.h

```diff
@@ -189,6 +193,163 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
   /// Override if the algorithm is not part of the Mantid distribution.
   const std::string helpURL() const override { return ""; }
 
+  template <typename T> void checkWorkspaceType() const {
+    static_assert(std::is_convertible<T *, MatrixWorkspace *>::value,
+                  "Workspace type needs to be convertible to MatrixWorkspace.");
+  }
+
+  inline bool isReserved(const std::vector<std::string> &list,
+                         const std::string &name) const {
+    return std::find(list.cbegin(), list.cend(), name) != list.cend();
+  }
+
+  using IPropertyManager::declareProperty;
+  using PropertyManagerOwner::declareProperty;
+  void declareProperty(std::unique_ptr<Kernel::Property> p,
+                       const std::string &doc = "") override;
+
+  Kernel::IPropertyManager::TypedValue
+  getProperty(const std::string &name) const override;
+
+  template <typename Workspace>
```

Comment:
This looks dangerous (or at least confusing), given that we also have a type named `Workspace`.